distributed/utils_test.py

```diff
@@ -651,98 +648,96 @@ def cluster(
         scheduler.daemon = True
         scheduler.start()
 
-        # Launch workers
-        workers = []
-        for i in range(nworkers):
-            q = mp_context.Queue()
-            fn = "_test_worker-%s" % uuid.uuid4()
-            kwargs = merge(
-                {
-                    "nthreads": 1,
-                    "local_directory": fn,
-                    "memory_limit": system.MEMORY_LIMIT,
-                },
-                worker_kwargs,
-            )
-            proc = mp_context.Process(
-                name="Dask cluster test: Worker",
-                target=_run_worker,
-                args=(q, scheduler_q, config),
-                kwargs=kwargs,
-            )
-            ws.add(proc)
-            workers.append({"proc": proc, "queue": q, "dir": fn})
+        with tempfile.TemporaryDirectory(prefix="_dask_test_worker") as tmpdirname:
+            # Launch workers
+            workers = []
+            for i in range(nworkers):
+                q = mp_context.Queue()
+                kwargs = merge(
+                    {
+                        "nthreads": 1,
+                        "local_directory": tmpdirname,
+                        "memory_limit": system.MEMORY_LIMIT,
+                    },
+                    worker_kwargs,
+                )
+                proc = mp_context.Process(
+                    name="Dask cluster test: Worker",
+                    target=_run_worker,
+                    args=(q, scheduler_q, config),
+                    kwargs=kwargs,
+                )
+                ws.add(proc)
+                workers.append({"proc": proc, "queue": q, "dir": tmpdirname})
 
-        for worker in workers:
-            worker["proc"].start()
-        try:
             for worker in workers:
-                worker["address"] = worker["queue"].get(timeout=5)
-        except queue.Empty:
-            pytest.xfail("Worker failed to start in test")
+                worker["proc"].start()
+            try:
+                for worker in workers:
+                    worker["address"] = worker["queue"].get(timeout=5)
+            except queue.Empty:
+                pytest.xfail("Worker failed to start in test")
```

Comment:
I think what's happening is the `proc` is not terminated unless all the workers start:

https://github.com/dask/distributed/blob/6fa676dec0b9be9c1802187dea56b7bc0aa94d8c/distributed/utils_test.py#L728

I think you'll need something like this?


```python
def close_queue(q):
    q.close()
    q._reader.close()
    q._write.close()

def terminate_process(proc):
    proc.terminate()
    proc.join(timeout=30)

with contextlib.ExitStack() as stack:
     for i in range(nworkers):
          tmpdirname = stack.enter_context(tempfile.TemporaryDirectory(prefix="_dask_test_worker"))
          q = mp_context.Queue()
          stack.callback(close_queue, q)

          proc = mp_context.Process(...)
          proc.start()
          stack.callback(terminate_process, proc)
```